scipy/special/specfun.c

```diff
@@ -0,0 +1,8426 @@
+
+    int k, m, mt;
+    double bs = 0.0, f = 0.0, f0 = 0.0, f1 = 1e-35;
+
+    for (m = 1; m < 901; m++) {
+        mt = (int)(0.5*log10(6.28*m)-m*log10(1.36*fabs(x)/m));
+        if (mt > 20) { break; }
+    }
+    if (m == 901) { m -= 1; }
+    for (k = m; k > -1; k--) {
+        f = 2.0*(k+1.0)*f1/x - f0;
+        if (k <= n) { bj[k] = f; }
+        if (k % 2 == 0) { bs += 2.0*f; }
+        f0 = f1;
+        f1 = f;
+    }
+    for (k = 0; k < (n+1); k++) {
+        bj[k] /= (bs - f);
+    }
+    dj[0] = -bj[1];
+    fj[0] = -bj[0] - dj[0]/x;
+    for (k = 1; k < (n+1); k++) {
+        dj[k] = bj[k-1] - k*bj[k]/x;
+        fj[k] = (k*k/(x*x)-1.0)*bj[k] - dj[k]/x;
+    }
+    return;
+}
+
+
+void specfun_cbk(int m, int n, double c, double cv, double qt, double *ck, double *bk) {
+    const double eps = 1.0e-14;
+
+    int i, i1, ip, j, k, n2, nm;
+    double r1, s1, sw, t;
+
+    ip = ((n - m) % 2 == 0 ? 0 : 1);
+    nm = 25 + (int)(0.5 * (n - m) + c);
+    double *u = calloc(200, sizeof(double));
+    double *v = calloc(200, sizeof(double));
+    double *w = calloc(200, sizeof(double));
+
+    u[0] = 0.0;
+    n2 = nm - 2;
+
+    for (j = 1; j < n2 + 1; j++) {
+        u[j] = c * c;
+    }
+    for (j = 1; j < n2 + 1; j++) {
+        v[j - 1] = (2.0 * j - 1.0 - ip) * (2.0 * (j - m) - ip) + m * (m - 1.0) - cv;
+    }
+    for (j = 1; j < nm; j++)
+        w[j - 1] = (2.0 * j - ip) * (2.0 * j + 1.0 - ip);
+
+    if (ip == 0) {
+        sw = 0.0;
+        for (k = 0; k < n2; k++) {
+            s1 = 0.0;
+            i1 = k - m + 1;
+
+            for (i = i1; i < nm + 1; i++) {
+                if (i < 0) { continue; }
+                r1 = 1.0;
+                for (j = 1; j <= k; j++) {
+                    r1 =  r1*(i + m - j) / (1.0*j);
+                }
+                s1 += ck[i] * (2.0 * i + m) * r1;
+                if (fabs(s1 - sw) < fabs(s1) * eps) { break; }
+                sw = s1;
+            }
+            bk[k] = qt * s1;
+        }
+    } else if (ip == 1) {
+        sw = 0.0;
+        for (k = 0; k < n2; k++) {
+            s1 = 0.0;
+            i1 = k - m + 1;
+
+            for (int i = i1; i < nm + 1; i++) {
+                if (i < 0) { continue; }
+                r1 = 1.0;
+                for (j = 1; j <= k; j++) {
+                    r1 = r1* (i + m - j) / (1.0*j);
+                }
+                if (i > 0) {
+                    s1 += ck[i - 1] * (2.0 * i + m - 1) * r1;
+                }
+                s1 -= ck[i] * (2.0 * i + m) * r1;
+
+                if (fabs(s1 - sw) < fabs(s1) * eps) { break; }
+                sw = s1;
+            }
+            bk[k] = qt * s1;
+        }
+    }
+
+    w[0] /= v[0];
+    bk[0] /= v[0];
+
+    for (k = 2; k <= n2; k++) {
+        t = v[k - 1] - w[k - 2] * u[k - 1];
+        w[k - 1] /= t;
+        bk[k - 1] = (bk[k - 1] - bk[k - 2] * u[k - 1]) / t;
+    }
+
+    for (k = n2 - 1; k >= 0; k--) {
+        bk[k - 1] -= w[k - 1] * bk[k];
+    }
```

Comment:
Here's what's causing the segfault. Writing to `bk[-1]`, and then you see the segfault when trying to free `bk` because the metadata about the allocated block got corrupted.

https://github.com/scipy/scipy/blob/6103deb58885562adcaee50307e0cef006d84a75/scipy/special/specfun/specfun.f#L576-L577

Reply:
Another facepalm moment.

Reply:
I missed it too and I spent tens of hours pouring over this code. On the whole, I think you’ve done a great job.